### ISSUE TITLE
Regression with parfor in numba=0.53.1 when aliased arrays are struct attributes

### ISSUE THREAD

**Issue by kozlov-alexey** [numba/numba#6960]
- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Hi, 

It seems in numba=0.53.x a new problem was introduced related to parfor's alias analysis, specifically `LoweringError `appears when compiled code uses an array attribute of some struct in several branches that further use this array in a prange. 

It can be reproduced with:
```
import numpy as np
from numba import njit
from numba.core import cgutils, types
from numba.extending import (
    make_attribute_wrapper,
    models,
    register_model,
    type_callable,
    lower_builtin,
)

def make_dummy_type():
    class Dummy(object):
        def __init__(self, data):
            self.data = data

    class DummyType(types.Type):
        def __init__(self, data):
            self.data = data
            super(DummyType, self).__init__(name="Dummy({})".format(self.data))

    @register_model(DummyType)
    class DummyModel(models.StructModel):
        def __init__(self, dmm, fe_type):
            members = [("data", fe_type.data)]
            models.StructModel.__init__(self, dmm, fe_type, members)

    make_attribute_wrapper(DummyType, "data", "_data")

    @type_callable(Dummy)
    def type_dummy(context):
        def typer(value):
            return DummyType(value)
        return typer

    @lower_builtin(Dummy, types.Array)
    def impl_dummy(context, builder, sig, args):
        typ = sig.return_type
        [value] = args
        dummy = cgutils.create_struct_proxy(typ)(context, builder)
        dummy.data = value
        return dummy._getvalue()

    # Store attributes
    return Dummy, DummyType

Dummy, DummyType = make_dummy_type()


''' This version fails with KeyError on size variable that's been removed'''
@njit(parallel=True)
def test_impl(arr, cond):

    dobj = Dummy(arr)
    if not cond:
        return dobj._data.min()

    return dobj._data.max()


''' And this doesn't and can be kind of workaround'''
@njit(parallel=True)
def test_impl_wa(arr, cond):

    dobj = Dummy(arr)
    dobj_data = dobj._data
    if not cond:
        return dobj_data.min()

    return dobj_data.max()

data = np.arange(20)
print(test_impl(data, True))
# print(test_impl_wa(data, True))
```

Gives:
> ...
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\parfors\parfor_lowering.py", line 69, in _lower_parfor_parallel
>     lowerer.lower_inst(instr)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 370, in lower_inst
>     val = self.lower_assign(ty, inst)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 544, in lower_assign
>     return self.lower_expr(ty, value)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 1070, in lower_expr
>     res = self.lower_call(resty, expr)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 806, in lower_call
>     res = self._lower_call_normal(fnty, expr, signature)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 1031, in _lower_call_normal
>     fnty, signature, expr.args, expr.vararg, expr.kws,
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 746, in fold_call_args
>     stararg_handler)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\typing\templates.py", line 239, in fold_arguments
>     ba.arguments[name] = normal_handler(i, param, ba.arguments[name])
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 729, in normal_handler
>     return self._cast_var(var, signature.args[index])
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 707, in _cast_var
>     val = self.loadvar(var.name)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 1292, in loadvar
>     ptr = self.getvar(name)
>   File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\lowering.py", line 1286, in getvar
>     return self.varmap[name]
> KeyError: '$26load_attr.1_size0.97'
> 
> During handling of the above exception, another exception occurred:
> ...
> numba.core.errors.LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
> '$26load_attr.1_size0.97'
> 
> File "numba\parfors\parfor.py", line 134:
>         def min_1(in_arr):
>             <source elided>
>             val = numba.cpython.builtins.get_type_max_value(in_arr.dtype)
>             for i in numba.parfors.parfor.internal_prange(len(in_arr)):
>             ^
> 
> During: lowering "id=1[LoopNest(index_variable = parfor_index.102, range = (0, $26load_attr.1_size0.97, 1))]{173: <ir.Block at 
> 

Debugging shows that it was first seen with #5936, and the root cause is probably in alias analysis and the fact that the shape and size definitions for aliased arrays are now inserted only once and it may turn up first in the block that won't directly use them, so despite other blocks still refer to them, the definitions are removed during simplify, i.e.

```
CFG adjacency lists:
{0: [14, 107],
 14: [24],
 24: [87, 89],
 87: [24],
 89: [],
 107: [171],
 171: [173, 191],
 173: [171],
 191: []}

-----------------------IR after array analysis: test_impl-----------------------
...
label 14:
    $26load_attr.1 = getattr(value=dobj, attr=_data) ['$26load_attr.1', 'dobj']
    $26load_attr.1_shape.96 = getattr(value=$26load_attr.1, attr=shape) ['$26load_attr.1', '$26load_attr.1_shape.96']
    $26load_attr.1_size0.97 = static_getitem(value=$26load_attr.1_shape.96, index=0, index_var=None) ['$26load_attr.1_shape.96', '$26load_attr.1_size0.97']
    $np_g_var.0 = global(np: <module 'numpy' from 'C:\\Users\\akozlov\\AppData\\Local\\Continuum\\anaconda3\\envs\\NUMBA_MASTER_4\\lib\\site-packages\\numpy\\__init__.py'>) ['$np_g_var.0']
...

--------------------------IR after simplify: test_impl--------------------------
...
label 14:
    $in_arr.48.107 = getattr(value=$dobj.106, attr=_data) ['$dobj.106', '$in_arr.48.107']
    $np_g_var.90 = global(np: <module 'numpy' from 'C:\\Users\\akozlov\\AppData\\Local\\Continuum\\anaconda3\\envs\\NUMBA_MASTER_4\\lib\\site-packages\\numpy\\__init__.py'>) ['$np_g_var.90']
    $np_typ_var.91 = const(str, int32)       ['$np_typ_var.91']
    $dtype_attr_var.92 = getattr(value=$np_g_var.90, attr=dtype) ['$dtype_attr_var.92', '$np_g_var.90']
    $dtype_var.93 = call $dtype_attr_var.92($np_typ_var.91, func=$dtype_attr_var.92, args=[Var($np_typ_var.91, parfor.py:133)], kws=(), vararg=None) ['$dtype_attr_var.92', '$dtype_var.93', '$np_typ_var.91']
    $60get_iter.8.75 = getiter(value=$58call_method.7.74) ['$58call_method.7.74', '$60get_iter.8.75']
---begin parfor 1---
index_var =  parfor_index.102
params =  None
races =  {'val.66'}
LoopNest(index_variable = parfor_index.102, range = (0, $26load_attr.1_size0.97, 1))
```

BR,
Alexey.

**Comment by stuartarchibald** [comment]
Thanks for the report. I can reproduce this locally, and reverting #5936 does "fix" it. The quotation marks around fix are because on both 0.52 and 0.53+revert of #5936 the reproducer segfaults?

CC @DrTodd13

**Comment by kozlov-alexey** [comment]
Yes, thanks, segfaulting script is my fault, I keep using buggy template without incref on stored value (corrected it now).

Also I found that actually removing single elif block in `_analyze_op_getattr`:
https://github.com/numba/numba/blob/287e94850f6a4faec38528cb4054a8298eaf3f75/numba/parfors/array_analysis.py#L1540-L1555
that seems to handle this specific case when expr is getattr and lhs type is array resolves it (tested on release0.53 branch) and other parfors/array_analysis tests also pass. Not sure whether this is a best fix at the moment, but at least it works.

**Comment by DrTodd13** [comment]
@kozlov-alexey There's another recent PR where I did something similar where shape and size calls were generated everywhere the array was assigned but used the same variables.  If you removed this elif and all the tests worked then I wonder why it was there in the first place.  I would explore generating shape calls generated in all branches.  Let me know if you need to discuss.
